operators/src/processing/point_in_polygon.rs

```diff
@@ -0,0 +1,775 @@
+
+        ensure!(
+            vector_sources[0].result_descriptor().data_type == VectorDataType::MultiPoint,
+            error::InvalidType {
+                expected: VectorDataType::MultiPoint.to_string(),
+                found: vector_sources[0].result_descriptor().data_type.to_string(),
+            }
+        );
+        ensure!(
+            vector_sources[1].result_descriptor().data_type == VectorDataType::MultiPolygon,
+            error::InvalidType {
+                expected: VectorDataType::MultiPolygon.to_string(),
+                found: vector_sources[1].result_descriptor().data_type.to_string(),
+            }
+        );
+
+        Ok(InitializedPointInPolygonFilter::new(
+            (),
+            vector_sources[0].result_descriptor(),
+            vec![],
+            vector_sources,
+            (),
+        )
+        .boxed())
+    }
+}
+
+pub type InitializedPointInPolygonFilter = InitializedOperatorImpl<(), VectorResultDescriptor, ()>;
+
+impl InitializedOperator<VectorResultDescriptor, TypedVectorQueryProcessor>
+    for InitializedPointInPolygonFilter
+{
+    fn query_processor(&self) -> Result<TypedVectorQueryProcessor> {
+        let point_processor = if let TypedVectorQueryProcessor::MultiPoint(point_processor) =
+            self.vector_sources[0].query_processor()?
+        {
+            point_processor
+        } else {
+            unreachable!("checked in `PointInPolygonFilter` constructor");
+        };
+
+        let polygon_processor = if let TypedVectorQueryProcessor::MultiPolygon(polygon_processor) =
+            self.vector_sources[1].query_processor()?
+        {
+            polygon_processor
+        } else {
+            unreachable!("checked in `PointInPolygonFilter` constructor");
+        };
+
+        Ok(TypedVectorQueryProcessor::MultiPoint(
+            PointInPolygonFilterProcessor::new(point_processor, polygon_processor).boxed(),
+        ))
+    }
+}
+
+pub struct PointInPolygonFilterProcessor {
+    points: Box<dyn VectorQueryProcessor<VectorType = MultiPointCollection>>,
+    polygons: Box<dyn VectorQueryProcessor<VectorType = MultiPolygonCollection>>,
+}
+
+impl PointInPolygonFilterProcessor {
+    pub fn new(
+        points: Box<dyn VectorQueryProcessor<VectorType = MultiPointCollection>>,
+        polygons: Box<dyn VectorQueryProcessor<VectorType = MultiPolygonCollection>>,
+    ) -> Self {
+        Self { points, polygons }
+    }
+
+    fn filter_points(
+        points: &MultiPointCollection,
+        polygons: &MultiPolygonCollection,
+        initial_filter: &BooleanArray,
+    ) -> Result<BooleanArray> {
+        let mut filter = Vec::with_capacity(points.len());
+
+        let tester = PointInPolygonTester::new(polygons);
+
+        let coordinates = points.coordinates();
+
+        for ((&coordinates_start_index, &coordinates_end_index), time_interval) in points
+            .multipoint_offsets()
+            .iter()
+            .tuple_windows()
```

Comment:
elegant solution but requires a new dependency and is more expensive than looping over an index and taking value at index+1 as end index. is it worth it?

Reply:
0be66e7